pkg/config/remote/client.go

```diff
@@ -15,96 +16,112 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/config/remote/data"
 	"github.com/DataDog/datadog-agent/pkg/config/remote/meta"
 	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
-	"github.com/DataDog/datadog-agent/pkg/remoteconfig/client"
-	agentgrpc "github.com/DataDog/datadog-agent/pkg/util/grpc"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/util/grpc"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
-	"github.com/DataDog/datadog-agent/pkg/version"
 	"github.com/pkg/errors"
-	"google.golang.org/grpc"
 	"google.golang.org/grpc/metadata"
 )
 
 // Client is a remote-configuration client to obtain configurations from the local API
 type Client struct {
-	m     sync.Mutex
-	ctx   context.Context
-	close func()
+	m sync.Mutex
+
+	ID string
+
+	startupSync sync.Once
+	ctx         context.Context
+	close       context.CancelFunc
 
 	agentName    string
+	agentVersion string
 	products     []string
-	pollInterval time.Duration
 
-	grpc           pbgo.AgentSecureClient
-	stateClient    *client.Client
-	currentConfigs client.Configs
+	pollInterval    time.Duration
+	lastUpdateError error
+
+	state *state.Repository
 
-	lastPollErr error
+	grpc pbgo.AgentSecureClient
 
-	apmSamplingUpdates chan []client.ConfigAPMSamling
-	cwsDDUpdates       chan []client.ConfigCWSDD
+	// Listeners
+	apmListeners []func(update map[string]state.APMSamplingConfig)
+	cwsListeners []func(update map[string]state.ConfigCWSDD)
 }
 
 // NewClient creates a new client
-func NewClient(agentName string, products []data.Product) (*Client, error) {
-	client, err := newClient(agentName, products)
+func NewClient(agentName string, agentVersion string, products []data.Product, pollInterval time.Duration) (*Client, error) {
+	ctx, close := context.WithCancel(context.Background())
+	grpcClient, err := grpc.GetDDAgentSecureClient(ctx)
 	if err != nil {
+		close()
 		return nil, err
 	}
-	go client.pollLoop()
-	return client, nil
-}
-
-func newClient(agentName string, products []data.Product, dialOpts ...grpc.DialOption) (*Client, error) {
-	ctx, close := context.WithCancel(context.Background())
-	grpcClient, err := agentgrpc.GetDDAgentSecureClient(ctx, dialOpts...)
+	repository, err := state.NewRepository(meta.RootsDirector().Last())
 	if err != nil {
 		close()
 		return nil, err
 	}
-	stateClient := client.NewClient(meta.RootsDirector().Last(), data.ProductListToString(products))
+
 	return &Client{
-		ctx:                ctx,
-		agentName:          agentName,
-		products:           data.ProductListToString(products),
-		grpc:               grpcClient,
-		close:              close,
-		pollInterval:       1 * time.Second,
-		stateClient:        stateClient,
-		apmSamplingUpdates: make(chan []client.ConfigAPMSamling, 8),
-		cwsDDUpdates:       make(chan []client.ConfigCWSDD, 8),
+		ID:           generateID(),
+		startupSync:  sync.Once{},
+		ctx:          ctx,
+		close:        close,
+		agentName:    agentName,
```

Comment:
❓ **Question**: Shouldn't we have `agentVersion` here as well ?

Reply:
Definitely either that or just remove it from the struct since I'm using `version.AgentVersion` when building the request (which is what would be passed in to New)!